topology/probes/k8s/container.go

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2017 Red Hat, Inc.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+package k8s
+
+import (
+	"sync"
+
+	"github.com/skydive-project/skydive/common"
+	"github.com/skydive-project/skydive/filters"
+	"github.com/skydive-project/skydive/logging"
+	"github.com/skydive-project/skydive/topology"
+	"github.com/skydive-project/skydive/topology/graph"
+
+	"k8s.io/api/core/v1"
+)
+
+type containerCache struct {
+	sync.RWMutex
+	defaultKubeCacheEventHandler
+	graph.DefaultGraphListener
+	*kubeCache
+	graph            *graph.Graph
+	podIndexer       *graph.MetadataIndexer
+	containerIndexer *graph.MetadataIndexer
+}
+
+// commonly accessed docker specific fields
+const (
+	DockerNameField         = "Docker.ContainerName"
+	DockerPodNamespaceField = "Docker.Labels.io.kubernetes.pod.namespace"
+	DockerPodNameField      = "Docker.Labels.io.kubernetes.pod.name"
+)
+
+func addNotNullFilter(m graph.Metadata, field string) {
+	m[field] = filters.NewNotFilter(filters.NewNullFilter(field))
+}
+
+func newContainerIndexer(g *graph.Graph) *graph.MetadataIndexer {
+	m := graph.Metadata{"Type": "container"}
+	addNotNullFilter(m, DockerPodNamespaceField)
+	addNotNullFilter(m, DockerPodNameField)
+
+	return graph.NewMetadataIndexer(g, m, DockerPodNamespaceField, DockerPodNameField)
+}
+
+func (c *containerCache) newMetadata(pod *v1.Pod, container *v1.Container) graph.Metadata {
+	m := newMetadata("container", container.Name, container)
+	common.SetField(m, DockerNameField, container.Name)
+	common.SetField(m, DockerPodNamespaceField, pod.GetNamespace())
+	common.SetField(m, DockerPodNameField, pod.GetName())
+	return m
+}
+
+func makeContainerUID(pod *v1.Pod, containerName string) string {
+	return string(pod.GetUID()) + "-" + containerName
```

Comment:
We should use a uuidv5 with pod UID and containerName instead of concatenating so that we will keep the UUID of graph element consistent in term of format.

Reply:
ok, have added a new function graph.NewIDNameBased(namespace, name string) which is now used to create a well-formed UID.